docs: Add field overrides how-to guide for panel plugins (#2754)
grafana/plugin-tools: docusaurus/docs/how-to-guides/panel-plugins/field-overrides.md

---
id: field-overrides
title: Add field override support to panel plugins
description: How to enable and customize field overrides in your panel plugin.
keywords:
  - grafana
  - plugins
  - plugin
  - panel
  - field overrides
  - field config
  - useFieldConfig
  - custom field config
---

# Add field override support to panel plugins

> Field overrides let dashboard users customize how individual fields are displayed in a panel. For example, a user might change the unit of a single series, apply a different color to one field, or set custom thresholds for a specific column in a table.

Reviewer: ```suggestion
Field overrides allow your users to customize how individual fields are displayed in a dashboard panel. For example, you can change the unit of a single series, apply a different color to one field, or set custom thresholds for a specific column in a table.
```

Author: Fixed in https://github.com/grafana/plugin-tools/pull/2754/changes/3db8b0247fd28c2ffcfcc68bc83332b95a986449 👍

This guide shows you how to enable field overrides in your panel plugin, customize which standard options are available, and define your own custom field configuration properties.

## Prerequisites

- A Grafana panel plugin with a `module.ts` entry point
- Familiarity with TypeScript and the `PanelPlugin` API

:::tip

> For a step-by-step guide to building a panel plugin from scratch, refer to the [panel plugin tutorial](../../tutorials/build-a-panel-plugin.md).

Reviewer: ```suggestion
For a step-by-step guide to building a panel plugin from scratch, refer to the [Build a panel plugin tutorial](../../tutorials/build-a-panel-plugin.md).
```

Author: Thanks, fixed in https://github.com/grafana/plugin-tools/pull/2754/changes/3db8b0247fd28c2ffcfcc68bc83332b95a986449 👍

        name: 'Show points',
        description: 'Whether to show data points on the line',
        defaultValue: defaultFieldConfig.showPoints,
      });
  },
});
```

The builder provides the following methods. All return the builder instance, so you can chain calls:

| Method               | Value type | Use case                            |
| -------------------- | ---------- | ----------------------------------- |
| `addNumberInput`     | `number`   | Free-form number input              |
| `addSliderInput`     | `number`   | Number input with a slider          |
| `addTextInput`       | `string`   | Free-form text input                |
| `addBooleanSwitch`   | `boolean`  | Toggle switch                       |
| `addSelect`          | `T`        | Dropdown select                     |
| `addRadio`           | `T`        | Radio button group                  |
| `addColorPicker`     | `string`   | Color picker                        |
| `addUnitPicker`      | `string`   | Unit picker dropdown                |
| `addFieldNamePicker` | `string`   | Field name picker                   |
| `addCustomEditor`    | `T`        | Fully custom React editor component |

> :::note

Reviewer: I'd do this as a level 3, not a note. E.g.

### Add a custom editor

Author: Good point, fixed in https://github.com/grafana/plugin-tools/pull/2754/changes/3db8b0247fd28c2ffcfcc68bc83332b95a986449 👍

        // Only visible when draw style is "line"
        showIf: (config) => config.drawStyle === 'line',
      });
  },
});
```

## Access field configuration in your panel component

Grafana applies both defaults and overrides to each field before passing data to your panel. The resolved configuration is available on each field through `field.config`, with custom properties under `field.config.custom`:

```tsx title="src/SimplePanel.tsx"
import React from 'react';
import { FieldType, PanelProps } from '@grafana/data';
import { Options, FieldConfig } from './types';

export const SimplePanel = ({ data, width, height }: PanelProps<Options>) => {
  const frame = data.series[0];

  return (
    <div style={{ width, height }}>
      {frame.fields
        .filter((field) => field.type === FieldType.number)
        .map((field, i) => {
          // Standard field config
          const unit = field.config.unit;
          const decimals = field.config.decimals;

          // Custom field config
          const custom = (field.config.custom ?? {}) as FieldConfig;
          const lineWidth = custom.lineWidth ?? 1;

          return (
            <span key={i}>

>               {field.display!(field.values[0]).text} ({unit}, {decimals} decimals, {lineWidth}px line)

Author: This is wrong, the `field.values` is a plain array since Grafana 10.
